### DISCUSSION TITLE
Does mobx makeAutoObservable() recurse?

### DISCUSSION THREAD

**Discussion by Lechros** [mobxjs/mobx#3250] [Q&A] [answered]
I'm using React, mobx, typescript, all newest version.
```
class Store {
  constructor() {
    makeAutoObservable(this);
  }
  people = new Map<string, Person>();
}

class Person {
  bag = new Bag()
}

class Bag {
  books = new Map<string, string>();
  pencils = [];
}

class Pencil {
  length = 2
}

const store = new Store();
let foo = new Person();
store.people.set("Tim", foo); // this should update
foo.bag.push(new Pencil()); // a. does this update?
foo.bag[0].length = 3; // b. does this update?
```
1. Is setting makeAutoObservable on Store enough to observe all children props? (line a and b)

```
store.people.get("Tim").bag.books = new Map<string, string>();
```
2. If I assign new Map like above, will mobx automatically convert it to ObservableMap? Do I have to convert to ObservableMap before assigning?

This is my first time using mobx, and couldn't figure out by reading documentation for makeAutoObservable. Thanks.

**Comment by mweststrate** [comment, marked as answer]
1. no, every class should mark it's own fields. 2. yes it will be converted automatically. 

TL,DR makeObservable is recursive, but only for built-ins: Map, Array, Set, and all objects that are not class instances (aka. plain objects; their prototype is `null` or `Object`)

**Reply by Lechros** [reply]
One more question, do I have to manually mark private methods as actions? Is it handled by makeAutoObservable by default?

**Reply by urugator** [reply]
https://mobx.js.org/api.html#introspection-utilities-
eg `console.log(isAction(this.privateMethod))`

**Reply by kubk** [reply]
> One more question, do I have to manually mark private methods as actions? Is it handled by makeAutoObservable by default?

TS private methods only exist in compile time. They are not different from public methods in runtime. And you can check it by yourself as urugator has suggested
